src/pkg/util/env/clean.go

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) 2018, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package env
+
+import (
+	"strings"
+
+	"github.com/opencontainers/runtime-tools/generate"
+	"github.com/singularityware/singularity/src/pkg/sylog"
+)
+
+// CleanEnv cleans environment variables before running the container
+func CleanEnv(g *generate.Generator, IsCleanEnv bool, environment []string) {
+
+	for _, env := range environment {
+		e := strings.SplitN(env, "=", 2)
+		if len(e) != 2 {
+			sylog.Verbosef("Can't process environment variable %s", env)
+			continue
+		}
+
+		// Transpose environment
+		if strings.HasPrefix(e[0], "SINGULARITYENV_") {
+			e[0] = strings.TrimPrefix(e[0], "SINGULARITYENV_")
+		} else if IsCleanEnv && (e[0] != "HOME" &&
+			e[0] != "TERM" &&
+			e[0] != "LANG" &&
+			e[0] != "http_proxy" &&
+			e[0] != "https_proxy" &&
+			e[0] != "no_proxy" &&
+			e[0] != "all_proxy" &&
+			e[0] != "ftp_proxy") {
+			continue
```

Comment:
I would include the SINGULARITY_NAME and SINGULARITY_CONTAINER variables here.  And then we must set them somewhere before this code executes.